contracts/fee/FeeHandler.sol

```diff
@@ -3,46 +3,215 @@
+                feeAmounts = _getFeeAmounts(1, feeAmount);
+            } else {
+                feeAmount = _getUint(Keys.claimableFeeAmountKey(markets[i], feeToken));
+                feeAmounts = _getFeeAmounts(2, feeAmount);
+            }
+
+            feeAmount = buybackToken == gmx ? feeAmounts.gmx : feeAmounts.wnt;
+            availableFeeAmount = availableFeeAmount + feeAmount;
+        }
+
+        uint256 maxFeeTokenAmount = _getMaxFeeTokenAmount(feeToken, buybackToken, batchSize);
+        if (availableFeeAmount >= maxFeeTokenAmount) {
+            return maxFeeTokenAmount;
+        } else {
+            return availableFeeAmount;
+        }
+    }
+
+    function _buybackFees(
+        address feeToken,
+        address buybackToken,
+        uint256 batchSize,
+        uint256 buybackAmount,
+        uint256 availableFeeAmount
+    ) private {
+        IERC20(feeToken).safeTransfer(msg.sender, buybackAmount);
+        IERC20(buybackToken).safeTransferFrom(msg.sender, address(this), batchSize);
+
+        _setAvailableFeeAmount(feeToken, buybackToken, availableFeeAmount - buybackAmount);
+    }
+
+    function _incrementAvailableFeeAmounts(uint256 version, address feeToken, uint256 feeAmount) private {
+        address wnt = dataStore.getAddress(Keys.WNT);
+
+        FeeAmounts memory feeAmounts = _getFeeAmounts(version, feeAmount);
+
+        uint256 availableFeeAmountGmx = _getAvailableFeeAmount(feeToken, gmx) + feeAmounts.gmx;
+        uint256 availableFeeAmountWnt = _getAvailableFeeAmount(feeToken, wnt) + feeAmounts.wnt;
+
+        _setAvailableFeeAmount(feeToken, gmx, availableFeeAmountGmx);
+        _setAvailableFeeAmount(feeToken, wnt, availableFeeAmountWnt);
+    }
+
+    function _setAvailableFeeAmount(address feeToken, address buybackToken, uint256 availableFeeAmount) private {
+        dataStore.setUint(Keys.buybackAvailableFeeAmountKey(feeToken, buybackToken), availableFeeAmount);
+    }
+
+    function _getAvailableFeeAmount(address feeToken, address buybackToken) private view returns (uint256) {
+        return _getUint(Keys.buybackAvailableFeeAmountKey(feeToken, buybackToken));
+    }
+
+    function _getFeeAmounts(uint256 version, uint256 feeAmount) private view returns (FeeAmounts memory) {
+        uint256 gmxFactor = _getUint(Keys.buybackGmxFactorKey(version));
+        FeeAmounts memory feeAmounts;
+
+        feeAmounts.gmx = Precision.applyFactor(feeAmount, gmxFactor);
+        feeAmounts.wnt = feeAmount - feeAmounts.gmx;
+        return feeAmounts;
+    }
+
+    function _getMaxFeeTokenAmount(
+        address feeToken,
+        address buybackToken,
+        uint256 batchSize
+    ) private view returns (uint256) {
+        uint256 feeTokenPrice = _getPriceFeedPrice(feeToken);
+        uint256 buybackTokenPrice = _getPriceFeedPrice(buybackToken);
+
+        uint256 expectedFeeTokenAmount = Precision.mulDiv(batchSize, feeTokenPrice, buybackTokenPrice);
```

Comment:
do i understand correctly, that buybackToken is GMX? in this case i guess it should be
```
Precision.mulDiv(batchSize, buybackTokenPrice, feeTokenPrice);
```

otherwise if batch size 100, GMX token price is $20 and fee token = USDC then expected fee token amount = 100 * 1 / 20 = 5 USDC. and i guess it should be 2000 USDC?

Reply:
i think you're right, this should be changed

Reply:
Yup, this was also identified in the audit as one of the findings and has been corrected in the latest commit